### PR TITLE
Convert proxy properties into URI proxy options for ice2 proxies

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR converts all proxy properties into URI proxy options. For example:
```
ice:hello?context=c1=TEST1,c2=TEST2,c3=TEST3,d%204=TEST%204&locator-cache-timeout=10s
```

Note: the proxy property `.Locator` has no equivalent proxy option.

This PR also updates to implementation to `Reference.ToString()` to _not_ print the following options where their value was explicitly set by the application (and therefore have the communicator's default value):
- InvocationTimeout
- LocatorCacheTimeout
- PreferNonSecure

Furthermore, IsRelative and CacheConnection are printed only when their value is different from the default (default for IsRelative is false, and default for CacheConnection is true).

This PR also renames IsConnectionCached/ConnectionCached to simply CacheConnection everywhere. `IsConnectionCached` sounds like it tells the caller whether the proxy holds a cached connection, when the actual semantics is whether or not this proxy caches the first connection it uses.

Finally this PR converts Context from a plain Dictionary<string, string> to a SortedDictionary<string, string>, which makes printed proxy contexts more readable.

### CODE REVIEW

File: csharp/src/Ice/Communicator-Properties.cs [hunk 1]
```diff
@@ -235,8 +235,10 @@ internal PropertyValue(string v, bool u)
             }
         }
 
-        /// <summary>Gets the value of a property as a proxy. If the property is not set, returns null.</summary>
-        /// <param name="name">The property name. The property name is also used as the prefix for proxy options.</param>
+        /// <summary>Gets the value of a property as a proxy. If the property is not set, returns null. For ice2
+        /// proxies, this method is equivalent to parsing the property's value.</summary>
+        /// <param name="name">The property name. For ice1 proxies, the property name is also used as the prefix for
```
Comment (bentoi) [inline csharp/src/Ice/Communicator-Properties.cs:240]: I would document the ice1 behavior in the summary.

**Review by bentoi — APPROVED** [review]
Looks good.

File: csharp/src/Ice/UriParser.cs [hunk 2]
```diff
@@ -333,6 +368,18 @@ internal static void RegisterCommon()
                 }
             }
             return (uri, altEndpoint, proxyOptions);
+
+            void CheckProxyOption(string name, bool alreadySet)
```
Comment (pepone) [inline csharp/src/Ice/UriParser.cs:372]: We can consider adding `pureEndpoints` and `uriString` and make this static

**Review by pepone — APPROVED** [review]
Looks good

File: csharp/src/Ice/Communicator-Properties.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/Communicator-Properties.cs:240]: Fixed.

File: csharp/src/Ice/UriParser.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/UriParser.cs:372]: I don't see the benefit of switching to a static local function here.

**Pull request merged**